python/ray/data/_internal/block_batching.py

```diff
@@ -61,19 +65,66 @@ def batch_blocks(
             select ``pandas.DataFrame`` or "pyarrow" to select
             ``pyarrow.Table``. Default is "native".
         drop_last: Whether to drop the last batch if it's incomplete.
+        shuffle: Whether to randomly shuffle the data using a local in-memory
+            shuffle buffer. This can only be used if a ``batch_size`` is specified.
+            This is a light-weight alternative to the global `.random_shuffle()`
+            operation; this shuffle will be less random but will be faster and less
+            resource-intensive.
+        shuffle_buffer_capacity: Soft maximum number of rows allowed in the local
+            in-memory shuffle buffer. This must be greater than or equal to
+            ``batch_size``. Note that this is a soft max: if the buffer is currently
+            smaller than this max, we will add a new data block to the buffer, but
+            this new data block may push the buffer over this max; we don't take the
+            size of the new data block into account when doing this capacity check.
+            Default is ``max(2 * shuffle_buffer_min_size, shuffle_buffer_min_size +
+            batch_size)`` if ``shuffle_buffer_min_size`` is given, otherwise the
+            default is ``10 * batch_size``.
+        shuffle_buffer_min_size: Minimum number of rows that must be in the local
+            in-memory shuffle buffer in order to yield a batch. This must be greater
+            than or equal to ``batch_size`` and must be less than
+            ``shuffle_buffer_capacity``. Increasing this will improve the randomness
+            of the shuffle but may increase the latency to the first batch.
+            Default is
+            ``max(min(shuffle_buffer_capacity // 2, shuffle_buffer_capacity -
+            batch_size), batch_size)``.
+        shuffle_seed: The seed to use for the local random shuffle.
 
     Returns:
         An iterator over record batches.
     """
-    batcher = Batcher(batch_size=batch_size)
-
-    def batch_block(block: ObjectRef[Block]):
-        with stats.iter_get_s.timer():
-            block = ray.get(block)
-        batcher.add(block)
+    if shuffle:
+        batcher = ShufflingBatcher(
+            batch_size,
+            shuffle_buffer_capacity,
+            shuffle_buffer_min_size,
+            shuffle_seed,
+        )
+    else:
+        batcher = Batcher(batch_size=batch_size)
+
+    def get_batches(block: Optional[ObjectRef[Block]] = None) -> Iterator[BatchType]:
+        if block is not None:
+            with stats.iter_get_s.timer():
+                block = ray.get(block)
+            # NOTE: Since we add one block at a time and then immediately consume
+            # batches, we don't check batcher.can_add() before adding the block.
+            batcher.add(block)
+        else:
+            batcher.done_adding()
         while batcher.has_batch():
+            # While the batcher has full batches, yield batches.
+            with stats.iter_next_batch_s.timer():
```

Comment:
Added a new iterator stage timer, since `next_batch()` can potentially be a bit expensive (building the shuffle buffer, generating random indices, etc.)